components/butterfly/src/rumor/dat_file.rs

```diff
@@ -49,24 +47,23 @@ pub struct DatFile {
 }
 
 impl DatFile {
-    pub fn new<T: AsRef<Path>>(member_id: &str, data_path: T) -> Self {
-        DatFile { path:        data_path.as_ref().join(format!("{}.rst", member_id)),
+    pub fn new(data_path: PathBuf) -> Self {
+        DatFile { path:        data_path,
                   header_size: 0,
                   header:      Header::default(), }
     }
 
     pub fn path(&self) -> &Path { &self.path }
 
-    pub fn read_into(&mut self, server: &Server) -> Result<()> {
-        let mut version = [0; 1];
-        let mut size_buf = [0; 8];
-        // JW: Resizing this buffer is terrible for performance, but it's the easiest way to
-        // read exactly N bytes from a file. I'm not sure what the right approach is but this
-        // won't be a performance issue for a long time anyway, if ever.
-        let mut rumor_buf: Vec<u8> = vec![];
-        let mut bytes_read = 0;
+    pub fn reader_for_file(&mut self) -> Result<BufReader<File>> {
```

Comment:
I think it would be safer and cleaner to store this reader within `DatFile` itself. It'll require a little re-jiggering since `DatFile::new` would need to return `Result<Self>` (perhaps implementing `TryFrom<PathBuf>` instead of `new` would make sense), but it removes a good amount of code and prevents consumers from doing something weird with the reader that would put `DatFile` in an inconsistent state.

Reply:
Ok, I'll give that a shot.

Reply:
This worked out great and ended up with much cleaner code than before.  Thanks for the suggestion!